pkg/mic/mic.go

```diff
@@ -500,91 +503,77 @@ func (c *Client) matchAssignedID(x *aadpodid.AzureAssignedIdentity, y *aadpodid.
-				// this assigned identity.
-				if oldAssignedID.Status.Status == aadpodid.AssignedIDCreated {
-					create = append(create, oldAssignedID)
-				}
-				break
-			}
+
+	for assignedIDName, newAssignedID := range new {
+		oldAssignedID, exists := old[assignedIDName]
+		idMatch := false
+		if exists {
+			idMatch = c.matchAssignedID(oldAssignedID, newAssignedID)
+		}
+		if idMatch && oldAssignedID.Status.Status == aadpodid.AssignedIDCreated {
+			// if the old assigned id is in created state, then the identity assignment to the node
+			// is not done. Adding to the list will ensure we retry identity assignment to node for
+			// this assigned identity.
+			glog.V(5).Infof("ok: %v, Create added: %s", idMatch, assignedIDName)
+			create[assignedIDName] = oldAssignedID
 		}
 		if !idMatch {
-			glog.V(5).Infof("ok: %v, Create added: %s", idMatch, newAssignedID.Name)
 			// We are done checking that this new id is not present in the old
 			// list. So we will add it to the create list.
-			create = append(create, newAssignedID)
+			glog.V(5).Infof("ok: %v, Create added: %s", idMatch, assignedIDName)
+			create[assignedIDName] = newAssignedID
 		}
 	}
 	stats.Put(stats.FindAssignedIDCreate, time.Since(begin))
-	return &create, nil
+	return create, nil
 }
 
-func (c *Client) getAzureAssignedIDsToDelete(old *[]aadpodid.AzureAssignedIdentity, new *[]aadpodid.AzureAssignedIdentity) (*[]aadpodid.AzureAssignedIdentity, error) {
+func (c *Client) getAzureAssignedIDsToDelete(old, new map[string]aadpodid.AzureAssignedIdentity) (map[string]aadpodid.AzureAssignedIdentity, error) {
+	delete := make(map[string]aadpodid.AzureAssignedIdentity)
 	// nothing to delete
-	if old == nil || len(*old) == 0 {
-		return nil, nil
+	if len(old) == 0 {
+		return delete, nil
 	}
 	// delete everything as nothing in new
-	if new == nil || len(*new) == 0 {
+	if len(new) == 0 {
 		return old, nil
 	}
 
-	delete := make([]aadpodid.AzureAssignedIdentity, 0)
-	var err error
-	idMatch := false
-
 	begin := time.Now()
-	for _, oldAssignedID := range *old {
-		idMatch = false
-		for _, newAssignedID := range *new {
-			idMatch, err = c.matchAssignedID(&newAssignedID, &oldAssignedID)
-			if err != nil {
-				glog.Error(err)
-				continue
-			}
-			//glog.Infof("Match %s %s %v", newAssignedID.Name, oldAssignedID.Name, idMatch)
-			if idMatch {
-				break
-			}
-		}
-		if !idMatch {
-			glog.V(5).Infof("ok: %v, Delete added: %s", idMatch, oldAssignedID.Name)
-			// We are done checking that this old id is not present in the new
-			// list. So we will add it to the delete list.
-			delete = append(delete, oldAssignedID)
+	for assignedIDName, oldAssignedID := range old {
+		newAssignedID, exists := new[assignedIDName]
+		idMatch := false
+		if exists {
+			idMatch = c.matchAssignedID(oldAssignedID, newAssignedID)
+		}
+		// assigned identity exists in the desired list too which means
+		// it should not be deleted
+		if exists && idMatch {
+			continue
```

Comment:
Do we want to simplify ?
if exists {
        if c.match... {
          continue
        }
}